drivers/fuel_gauge/bq27z746/emul_bq27z746.c

```diff
@@ -0,0 +1,352 @@
+		return -EIO;
+	}
+
+	memset(buf, 0, len);
+
+	/*
+	 * The data read from BQ27Z746_ALTMANUFACTURERACCESS is:
+	 * 0..1: The command (for verification)
+	 * 2..33: The data
+	 * 34: Checksum calculated as (uint8_t)(0xFF - (sum of all command and data bytes))
+	 * 35: Length including command, checksum and length (e.g. data length + 4)
+	 */
+
+	/* Put the command in the first two byte */
+	sys_put_le16(data->mac_cmd, buf);
+
+	/* Based on the command, put some data and the length into the buffer. */
+	/* In all of the operations, don't consider the zero-terminator. */
+	switch (data->mac_cmd) {
+	case BQ27Z746_MAC_CMD_MANUFACTURER_NAME:
+		memcpy(&buf[2], manufacturer_name, sizeof(manufacturer_name) - 1);
+		buf[35] = sizeof(manufacturer_name) - 1 + BQ27Z746_MAC_OVERHEAD_LEN;
+		break;
+	case BQ27Z746_MAC_CMD_DEVICE_NAME:
+		memcpy(&buf[2], device_name, sizeof(device_name) - 1);
+		buf[35] = sizeof(device_name) - 1 + BQ27Z746_MAC_OVERHEAD_LEN;
+		break;
+	case BQ27Z746_MAC_CMD_DEVICE_CHEM:
+		memcpy(&buf[2], device_chemistry, sizeof(device_chemistry) - 1);
+		buf[35] = sizeof(device_chemistry) - 1 + BQ27Z746_MAC_OVERHEAD_LEN;
+		break;
+	default:
+		LOG_ERR("ALTMAC command 0x%x is not supported", data->mac_cmd);
+		return -EIO;
+	}
+
+	/* Calculate the checksum */
+	uint8_t sum = 0; /* Intentionally 8 bit wide and overflowing */
+
+	for (int i = 0; i < BQ27Z746_MAC_COMPLETE_LEN - 2; i++) {
+		sum += buf[i];
+	}
+	buf[34] = 0xFF - sum;
+
+	return 0;
+}
+
+static int emul_bq27z746_write(const struct emul *target, uint8_t *buf, size_t len)
+{
+	struct bq27z746_emul_data *data = target->data;
+	const uint8_t reg = buf[0];
+
+	switch (reg) {
+	case BQ27Z746_ALTMANUFACTURERACCESS:
+		data->mac_cmd = sys_get_le16(&buf[1]);
+		return 0;
+	default:
+		LOG_ERR("Writing is only supported to ALTMAC currently");
+		return -EIO;
+	}
+}
+
+static int emul_bq27z746_reg_read(const struct emul *target, int reg, int *val)
+{
+	switch (reg) {
+	case BQ27Z746_MANUFACTURERACCESS:
+		*val = 1;
+		break;
+	case BQ27Z746_ATRATE:
+		*val = -1;
+		break;
+	case BQ27Z746_ATRATETIMETOEMPTY:
+		*val = 1;
+		break;
+	case BQ27Z746_TEMPERATURE:
+		*val = 1;
+		break;
+	case BQ27Z746_VOLTAGE:
+		*val = 1;
+		break;
+	case BQ27Z746_BATTERYSTATUS:
+		*val = 1;
+		break;
+	case BQ27Z746_CURRENT:
+		*val = -1;
+		break;
+	case BQ27Z746_REMAININGCAPACITY:
+		*val = 1;
+		break;
+	case BQ27Z746_FULLCHARGECAPACITY:
+		*val = 1;
+		break;
+	case BQ27Z746_AVERAGECURRENT:
+		*val = -1;
+		break;
+	case BQ27Z746_AVERAGETIMETOEMPTY:
+		*val = 1;
+		break;
+	case BQ27Z746_AVERAGETIMETOFULL:
+		*val = 1;
+		break;
+	case BQ27Z746_MAXLOADCURRENT:
+		*val = 1;
+		break;
+	case BQ27Z746_MAXLOADTIMETOEMPTY:
+		*val = 1;
+		break;
+	case BQ27Z746_AVERAGEPOWER:
+		*val = 1;
+		break;
+	case BQ27Z746_BTPDISCHARGESET:
+		*val = 1;
+		break;
+	case BQ27Z746_BTPCHARGESET:
+		*val = 1;
+		break;
+	case BQ27Z746_INTERNALTEMPERATURE:
+		*val = 1;
+		break;
+	case BQ27Z746_CYCLECOUNT:
+		*val = 100;
```

Comment:
I think I missed this earlier.

This is set to `100` but the test verifies `1`. Do we have an endianness problem?

Reply:
If it's the emulator, let me know if you want any help and I'll download the PR and take a look.

Reply:
Maybe I misunderstood the API? It says:

```
	/** Cycle count in 1/100ths (number of charge/discharge cycles) */
	FUEL_GAUGE_CYCLE_COUNT,
```

So in my understanding 100 actual cycles should show up as 1 at the sensor API.

Reply:
Nope your understanding is correct! My apologies! I had forgotten about the `1/100`ths.